### PR TITLE
[FLINK-3648] Introduce Trigger Test Harness

### PR DESCRIPTION
**Pull request by kl0u**
Adds a test harness that facilitates the testing of custom triggers and window assigners.

### CODE REVIEW

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/util/WindowingTestHarness.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+	private final ConcurrentLinkedQueue<Object> expectedOutputs = new ConcurrentLinkedQueue<>();
+
+	private volatile boolean isOpen = false;
+
+	public WindowingTestHarness(ExecutionConfig executionConfig,
+								WindowAssigner<? super IN, W> windowAssigner,
+								TypeInformation<K> keyType,
+								TypeInformation<IN> inputType,
+								TypeSerializer<W> windowSerializer,
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/util/WindowingTestHarness.java]: This should be retrievable from the `WindowAssigner`.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/util/WindowingTestHarness.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+								KeySelector<IN, K> keySelector,
+								Trigger<? super IN, ? super W> trigger,
+								long allowedLateness) {
+
+		ListStateDescriptor<IN> windowStateDesc = new ListStateDescriptor<>("window-contents",
+			inputType.createSerializer(new ExecutionConfig()));
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/util/WindowingTestHarness.java]: We could use the passed in `ExecutionConfig` here.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/util/WindowingTestHarness.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+	 * Adds an element to the expected output.
+	 * The expected output should contain the elements and watermarks that we expect the output of the operator to
+	 * contain, in the correct order. This will be used to check if the produced output is the expected one, and
+	 * thus determine the success or failure of the test.
+	 */
+	public void addExpectedElement(IN element, long timestamp) {
+		expectedOutputs.add(new StreamRecord<>(element, timestamp));
+	}
+
+	/**
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/util/WindowingTestHarness.java]: We should maybe mention that this internally uses an assert and the test thus fails when the expected output does not match the actual output.

**Comment by aljoscha** [comment]
Looks good! I just had some minor comments. After fixing those it should be good to merge.

**Comment by kl0u** [comment]
Thanks for the comments @aljoscha !
I integrated them.

**Comment by aljoscha** [comment]
Thanks for your work. 😃 I merged it. Could you please take care of closing this PR and the issue?

**Pull request closed without merging**